### ISSUE TITLE
Make our HAL+JSON more HAL compliant

### ISSUE THREAD

**Issue by amitaibu** [RESTful-Drupal/restful#186]
http://stateless.co/hal_specification.html

Seriously @mirzu, this issue has your name all over it! :)

**Comment by amitaibu** [comment]
@mateu-aguilo-bosch @mirzu do we need to start by removing the "data" wrapper from the HAL-JSON formatter plugin?

![local_d7_dev_api_v1_articles_](https://cloud.githubusercontent.com/assets/125707/4669346/6d32119c-556d-11e4-94a7-5ad3e5b84bf7.jpg)

(See the ng-admin [issue](https://github.com/marmelab/ng-admin/issues/86) that made me think of it)

**Comment by amitaibu** [comment]
Ignore above comment, seems wrapping with `data` is a common practice.

**Comment by mirzu** [comment]
I'm curious to see other examples where data is used to list embedded resources. 

From the specification it looks like what we are wrapping in data belongs in the reserved property "_embedded" The reason for using _embedded is that API client libraries will treat _embedded as a list of resources that can potentially be followed, rendered etc,  while they will treat data a property.

 The reserved "_embedded" property is OPTIONAL

   It is an object whose property names are link relation types (as
   defined by [RFC5988]) and values are either a Resource Object or an
   array of Resource Objects.

   Embedded Resources MAY be a full, partial, or inconsistent version of
   the representation served from the target URI.

**Comment by mirzu** [comment]
More thoughts:
The way we are doing it isn't invalid, but using embedded for lists of resources would be better.

Wrapping the properties in a data object isn't bad, but I'm curious to know what we gain by doing so. 

Wrapping _embdedded in data along with properties would be bad.

**Comment by amitaibu** [comment]
> I'm curious to see other examples where data is used to list embedded resources.

@mirzu , the examples I've seen were not HAL specific. But in my case, the fact that we wrap our data with `data` was very handy, as I just had to point ng-admin to the correct key, rather then ignoring other root level keys.

**Comment by mirzu** [comment]
One reason to make the API follow HAL conventions more closely is this little gem:

It's a HAL specific ng-admin type thing that is 100% drop in 0 config if you implement enough of the features. 
Example that is on it's way: http://mirzu.fourword.webchefs.org/hal-browser/browser.html#/api/v1/blogposts/142
Source: https://github.com/mikekelly/hal-browser
Full Example: http://haltalk.herokuapp.com/explorer/browser.html#/

**Comment by amitaibu** [comment]
> Example that is on it's way: http://mirzu.fourword.webchefs.org/hal-browser/browser.html#/api/v1/blogposts/142

Seems to be password protected ;)

hal-browser looks cool. #200 is probably to help there as-well.

**Comment by mirzu** [comment]
Shouldn't need a username and password anymore.

On Mon, Oct 20, 2014 at 2:29 PM, Amitai Burstein notifications@github.com
wrote:

> Example that is on it's way:
> http://mirzu.fourword.webchefs.org/hal-browser/browser.html#/api/v1/blogposts/142
> 
> Seems to be password protected ;)
> 
> hal-browser looks cool. #200 https://github.com/Gizra/restful/pull/200
> is probably to help there as-well.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Gizra/restful/issues/186#issuecomment-59825399.

## 

Michal Minecki
DevOps Track Lead | Drupalcon Austin 2014
Director of Technology | Four Kitchens
  | mike@fourkitchens.com
  | +1 512 870 8441 (direct)
  | +1 562 551 8775 (cell)

Four Kitchens:  We make BIG websites
  | http://www.fourkitchens.com
  | +1 512 454 6659

**Comment by e0ipso** [comment]
I see [here](https://github.com/mikekelly/hal-browser#about-hal) that the `self` property should be in the `_links` section (which makes a lot of sense). I created https://github.com/Gizra/restful/issues/219 for that.

**Comment by amitaibu** [comment]
HAL advocates for discovery and easy learning of the API while using the API. We have improved the discovery, and need to think about ways to inject human readable hints to the response.

Now I'd like to better understand why HAL is better than what we have. 

Here's an example of our current structure for `api/articles`, assuming we have a single article that references a single page.

Existing:

``` javascript
{
  data: [
    {
      label: 'This is the Article',
      // Self link of the parent.
      self:  '/api/articiles/1'
      reference: {
        label: 'This is the Page',
          // Self link of the child.
          self: "/api/pages/2" 
      }
    }
  ],
  _links: {
    self: { href: '/api/articiles'},
    next: { href: '/api/articiles?page=2'}
  }
}
```

And in HAL it should be if I understand it correctly, which I might not:

``` javascript
{
  data: [
    {
      label: 'This is the Article',
      _links: {
        self: { href: '/api/articiles/1'}
      }
      _embedded: {
        ea:reference: [{
            label: 'This is the Page',
            _links: {
              self: { "href": "/api/pages/2" },
            },
        },
      }
    }
  ],
  _links: {
    self: { href: '/api/articiles'},
    next: { href: '/api/articiles?page=2'}
    curies: [
      {
        name: "ea",
        href: "com/docs/rels/{rel}",
        templated: true
      }
    ],
  }
}
```

Personally, I don't think the HAL implementation is easier to read -- took me time to read the spec then just to figure our existing example, but my concern is that in the HAL structure it's not clear that I can create/ update the resource and it's embedded resources in one single call.

In short - can someone please explain, with _real life_ example, why HAL is the right way? :)

**Comment by e0ipso** [comment]
> In short - can someone please explain, with real life example, why HAL is the right way? :)

As with everything it's debatable what's best or worst. You have to balance the goods and bads to this as well.

HAL allows you to conserve bandwidth by optionally embedding link payloads in the original request. HAL also provides a scalable way of organizing complex data structures.

> it's not clear that I can create/ update the resource and it's embedded resources in one single call.

Even thought that is super nice, I don't believe it's a common practise. That's why we need to be explicit about it in the docs. Having the current simplified structure does not convey that either, it's just that we know that we can; although I grant you that it's easier to grasp the data structure you need to send.

I'm not decided for any side here, but I definitely want to hear @mirzu's thoughts.

**Comment by mirzu** [comment]
I've been thinking about this all morning and trying to make sure I'm clear in my own head.

1.) This is somewhat arbitrary, and interoperability should trump arbitrary. Therefore we should stick with the standard, and if that standard is HAL we should embrace it wholeheartedly, write a new standard, or attempt to change the standard. 

The biggest takeaway from all the API design talks and books I've been going through is **"don't be unique."** Building things that are interchangeable in theory makes it easier for others to build on each others work. My biggest reason for pushing for HAL is that it's somewhat popular and is going to be the basis for D8, not because it's beautiful. It's not. It's just good enough, and popular. If it sucked and was popular I'd say ditch it. 

**Perhaps before we ask the question is x more elegant than y we should ask is x more interoperable than y.** 

2.) As for the merits of either format. In your example there is no difference between _label_ and _reference_ They are both just properties. In the HAL example it's very explicit that reference and label are and referenced object and a property, therefore making it easier to write a client that follows or expands references and displays properties regardless of their names. 

3.) The curries make semi-automatic docs easy, but do make it harder for humans to read. The organization of the embedded resources is harder to grok in the HAL example, but it's easier to abstract.

4.) Striving for parity between GET format and expected POST format is helpful, but I would place more emphasis on the above. As @mateu-aguilo-bosch mentioned we are going to need docs, and I imagine that there will be tons of uses cases that will need to break that parity making it impossible to build abstractions that expect the parity.

**Comment by amitaibu** [comment]
Thanks for the answers folks! 

I agree that choosing a "good enough" yet popular standard is the way to go. Since RESTful has been challenging quite a few of the concepts other Drupal modules had, I wanted to make sure we were not jumping to HAL without paying careful attention to it.

Now for the technical part - we need to figure out how to add HAL integration without necessarily forcing it on other formats.

**Comment by e0ipso** [comment]
> Now for the technical part - we need to figure out how to add HAL integration without necessarily forcing it on other formats.

I was thinking about adding some sort of generic metadata that is offered to the formatters to use. Maybe similar to how we're handling the `$request` "object". Just dump an structured array and make the necessary assumptions when reading it in the formatter. We can be more structured later if we see the point.

**Comment by amitaibu** [comment]
> I was thinking about adding some sort of generic metadata that is offered to the formatters to use. Maybe similar to how we're handling the $request "object".

Not sure I get what you mean... :)

**Comment by mirzu** [comment]
I was thinking the same thing!

$data would be an object that had properties, instead of just a hash. I
think there is a generic subset of classifications that we can use as an
abstraction that the different formats all share. I think they are:
$properties -> nakid properties in hal
$links -> _link
$embedded -> _embedded

Which nicely corresponds to HAL. I just looked at Siren:
$properties -> properties
$embedded -> entities
$links -> links

Siren also has 'actions' and a 'class'

I'll look at Collection+json and json-LD

to see if this assumption makes sense.

On Wed, Oct 29, 2014 at 1:36 PM, Amitai Burstein notifications@github.com
wrote:

> I was thinking about adding some sort of generic metadata that is offered
> to the formatters to use. Maybe similar to how we're handling the $request
> "object".
> 
> Not sure I get what you mean... :)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Gizra/restful/issues/186#issuecomment-60980139.

## 

Michal Minecki
DevOps Track Lead | Drupalcon Austin 2014
Director of Technology | Four Kitchens
  | mike@fourkitchens.com
  | +1 512 870 8441 (direct)
  | +1 562 551 8775 (cell)

Four Kitchens:  We make BIG websites
  | http://www.fourkitchens.com
  | +1 512 454 6659

**Comment by mirzu** [comment]
collection+json: 
$properties -> attributes
$links -> links
$embedded -> collection?

From a quick read through the collection+json is really designed to work as always having a list list of items so things are always "embedded."

JSON-LD 
um no. Just no. So complicated, so abstract that it's not really worth exploring.

**Comment by amitaibu** [comment]
Ok, I have a suggestion for approach:
1. On the formatter handler we will add a prepareRow() method, that will process a single result.
2. Since there are multiple implementations it is going to be hard (as in impossible) to keep the `publidFieldsInfo()` deceleration completely formatter agnostic. 
   For `HAL` Things like `_embedded` could be automatically taken if the public field has a `resource` property.

But for stuff like the self, or curies we need to be format specific. So I was thinking something like:

``` php
  /**
   * Overrides RestfulExampleArticlesResource::publicFieldsInfo().
   */
  public function publicFieldsInfo() {
    $public_fields = parent::publicFieldsInfo();

    $public_fields['self'] = array(
      'callback' => array($this, 'getEntitySelf'),
      'formatter_options' => array(
        'hal_json' => array(
          // We are able to add some metadata here.
          'links' => TRUE,
          'type' => 'href'
        ),
      ),
    ); 
  }
```

**Comment by e0ipso** [comment]
@amitaibu, @mirzu I think that the formatter options belong to the formatter plugin. I feel that this is getting a bit more complicated than it needs to be. For me, the principle is **restful**: _we want to be able generate a structured array with data to send to the consumer_, **formatter**: _can take an structured array and format it in the desired format_. I don't see the intrinsic dependency between the both.

That being said, I think that something like the `prepareRow` suggestion is actually pretty clean. That could be called from the `prepare` method. Also, since we have the `isList` method we can adapt to the problems for the collections.

All in all I would like to have the `hal_json` formatter unset the `'self'` property to put it inside the `_links` section (during the `prepare`).

**Comment by amitaibu** [comment]
how would you deal with curies?

**Comment by e0ipso** [comment]
Curies are information about the different related resources. We already have that available to us (introduced in the discovery effort). We can bring that information in and format it as curies (adding the missing bits first if necessary).

Ultimately it will boil down to add some metadata to the field, as you suggest, but I'd rather have meaningful information that can be reused by several subsystems than have many metadata keys in the field definition.

**Comment by amitaibu** [comment]
To answer https://github.com/Gizra/restful/pull/230#issuecomment-61617237

> Also, I'm having a travel ordeal on Saturday and I'd like to have restful stuff to do. Can we prioritize?

Seems to me that trying to tackle this one should help you get your Saturday travels productive :P

**Comment by mirzu** [comment]
restful: we want to be able generate a structured array with data to send to the consumer 
I think our lives will be much easier in the **formatter** if our structured array has some structure. 

That structure can come from the above suggestion or it can be an object in which each item has properties for some maybe all of the stuff in the publicFieldsInfo and info about the processed field. 

There are lots of subtleties that we are enforcing that become difficult to modify later and require weird inspection of stuff.

I think either approach has merit, but the current unstructured array has some limitations.

**Comment by mirzu** [comment]
So my "hack" to make an example that is more HAL compliant is here:
http://mirzu.fourword.webchefs.org:8080/api/v1/blogposts
HAL Browser here: 
http://mirzu.fourword.webchefs.org:8080/hal-browser/browser.html#/api

(Use :8080, :80 will give you a login link)

The changes I have made to the restful module are here: 
https://github.com/fourkitchens/restful/pull/1/files

The example module that uses the changes I made is here:
https://github.com/fourkitchens/restful_fourword

The changes are minor, and I didn't make them in a way that is scalable or that should be included in the module because my goal was to really see if there is any point in being a stickler on some of these things. Thats the focus of Matt and I's talk at Badcamp. After going through the exercise I can confidently say that HATEOS has advantages, and being strict with HAL has advantages, but there are things that "feel" weird coming from a Drupal perspective and make the API response harder to Grok for people that are not familiar with Hypermedia APIs.

Here are issues for the things I found:
fix _links: #235
fix referenced resources: #236
fix /api links: #237

**I think we are at a fork in the road with HAL.**

Either we make it strict or make a new default formatter that doesn't call it's self HAL. That new formatter can still be a HATEOS API without being HAL. I can understand why this might be desirable, but it has some disadvantages.

My vote is to keep the default formatter HAL, make our HAL implementation compliant, and ship with the hal-browser, and perhaps even write tests that insure that the HAL browser works as expected. It's a lot easier to understand the format and the subtleties with the client, at least it was for me. I imagine that would also make it a great way for new users to learn how things work. 

We could then extend the HAL browser slightly so that it can understand our OPTIONS links and create templates for the POST dialog box it includes. 

![the_hal_browser](https://cloud.githubusercontent.com/assets/251658/4921388/1809f17e-650a-11e4-9831-eba7e5fc8591.png)

**Comment by amitaibu** [comment]
> My vote is to keep the default formatter HAL, make our HAL implementation compliant,

I agree with this, for the following reasons:
As @mirzu stated, this is becoming a standard, and is already adopted by D8.
But I think that RESTful is in a point that it doesn't only follow best practices, but in fact helps shape them. So, if RESTful will ship with a strict HAL formatter, people will be "forced" to follow the best practice.
The `RestfulFormatterBase` could be the same of what we call today HAL (which is basicaly a clean json with very little HATEOS in the `_links`).

> and perhaps even write tests that insure that the HAL browser works as expected.

Luckily, tests are a prerequisite for getting HAL - and to that matter any feature - in :)

I think that if @mateu-aguilo-bosch also gives us a green light we should move forward and get some code to discuss.

**Comment by amitaibu** [comment]
Admittedly I have second thoughts about making HAL the _default_ formatter.

I would like to be able to use HAL, but for new comers / ppl who just want to get a clean JSON (as actually I currently do in all my project), having a clean JSON + with little metadata would look more appealing.

**Comment by amitaibu** [comment]
> Admittedly I have second thoughts about making HAL the default formatter.

Had some time to think about why HAL should be added, but _not_ as default formatter:
1. Are there examples of big sites, such as Github, flickr, twitter that already HAL? No.
2. Do you have a _real_ use case where HAL would _greatly_ be better than our "simple" json? I personally don't
3. Would client side be easier to work with the received data, and would it be easier for front end developers to understand HAL vs the "simple" json? I don't think so.

So why do I still think HAL should be added?
1. RESTful should provide, for advanced developers, the possibility to use this standard. The fact that D8 chose it is also important, although lets not forget that Drupal had xmlrpc as-well :P
2. It would help improve and validate RESTful's API robustness
3. It will get geeky developers excited, and they will contribute to the project ;)

**Comment by amitaibu** [comment]
See #238

**Comment by e0ipso** [comment]
@amitaibu good arguments, you made me switch my mind. I just want to make sure HAL is not buried somewhere that no one uses.

Maybe it's time for a UI that exposes all these variables and give a preponderant place for this one?

@amitaibu, @mirzu thoughts?

**Comment by amitaibu** [comment]
Cool, I'm ok with a settings page

‫ב-12 בנוב 2014, בשעה 17:14, ‏Mateu Aguiló Bosch notifications@github.com כתב/ה:‬

**Comment by mirzu** [comment]
Hi Mateo and Amitai I want to introduce @patrickocoffeyo a Web Chef who luckily mentioned he wanted to find awesome open source projects to help out on. I can't think of a better project I've mentioned him in a couple of issues that I think might be good ways to toe in. . .

**Comment by e0ipso** [comment]
I'm so sorry I missed this message while I was away. Please, come in Patrick and have fun!

**Comment by e0ipso** [comment]
Also, I think that after merging #256, we can close this one.
